stats.rst

```diff
@@ -0,0 +1,271 @@
+
+^^^^^^
+Docker
+^^^^^^
+
+.. code-block:: console
+
+  $ docker stop graphite
+
+  # find the configuration directory
+  # listed as "Source"
+  $ docker inspect graphite | jq '.[].Mounts[] | select(.Destination == "/opt/graphite/conf")'
+  {
+    "Type": "volume",
+    "Name": "73025f5bcf4c9d761be5eec05cd8ed4169ef0028569ce130d84b69ce46f6d7c8",
+    "Source": "/var/lib/docker/volumes/73025f5bcf4c9d761be5eec05cd8ed4169ef0028569ce130d84b69ce46f6d7c8/_data",
+    "Destination": "/opt/graphite/conf",
+    "Driver": "local",
+    "Mode": "",
+    "RW": true,
+    "Propagation": ""
+  }
+
+  $ sudo vim $Source/storage-schemas.conf
+  # comment out this default
+  #[default_1min_for_1day]
+  #pattern = .*
+  #retentions = 60s:1d
+
+  # catch all metrics that start with 'flux'
+  # and retain each data point as 1s
+  [flux]
+  pattern = ^flux.*
+  retentions = 1s:3d
+
+  # remove the old whisper files
+  $ cd $(docker inspect graphite | \
+    jq '.[].Mounts[] | select(.Destination == "/opt/graphite/storage") | .Source' && \
+    sudo rm -rf whisper/*
+
+  # restart the docker image and the new configuration should take effect
+  $ docker start graphite
+
+^^^^^^^^^^^^^^^
+Packages/Source
+^^^^^^^^^^^^^^^
+
+First make sure that the carbon cache is enable on boot.
+
+.. code-block:: console
+
+  $ sudo vim /etc/default/graphite-carbon
+  # set to true
+  CARBON_CACHE_ENABLED=true
+
+Next, you will need to edit carbon's config file and set the specific you want. You can fin specifics on the configurations `here <https://graphite.readthedocs.io/en/stable/config-carbon.html>`_.
+The carbon config file is stored in the ``CONF_DIR`` that was found earlier.
+
+.. code-block:: console
+
+  $ sudo vim /etc/carbon/storage-schemas.conf
+  # comment out this default
+  #[default_1min_for_1day]
+  #pattern = .*
+  #retentions = 60s:1d
+
+  # catch all metrics that start with 'flux'
+  # and retain each data point as 1s
+  [flux]
+  pattern = ^flux.*
+  retentions = 1s:3d
+
+Then you will need to clear out any existing whisper files in the database which is store in the ``LOCAL_DATA_DIR`` found earlier.
+
+.. code-block:: console
+
+  $ sudo rm -rf /val/lib/graphite/whisper/*
+
+Now restart the carbon daemon and the new configuration should take effect.
+
+.. code-block:: console
+
+  $ sudo systemctl restart carbon-cache
+
+--------------------------------
+Running Stats Collection in Flux
+--------------------------------
+
+Once you have brubeck and Graphite setup and Graphite running there are only two more things left to be able to get stats from Flux.
+First you will need to set the ``FLUX_FRIPP_STATSD`` environment variable with the endpoint of the aggregator (brubeck in this case).
+If you just used the default config for brubeck as suggested earlier, the endpoint is ``0.0.0.0:8126``. If you used the docker setup the
+endpoint is ``0.0.0.0:8125``.
```

Comment:
Using `0.0.0.0` (INADDR_ANY) is a bit weird as a `sendto()` address.  Usually one uses this to bind a service to "all interfaces".  I'm not really sure what it means in this context, but it does seem to work!

One vote for `127.0.0.1` here, as "less wierd".  It's possible I'm ignorant about how that address works though!

Reply:
from the default brubeck config:

```bash
"samplers" : [
    {
      "type" : "statsd",
      "address" : "0.0.0.0",
      "port" : 8126,
      "workers" : 4,
      "multisock" : true,
      "multimsg" : 8
    }
  ]
```

Not sure if it makes a difference  to use `127.0.0.1`, but I was just using the config as a reference.

Reply:
`0.0.0.0` is the INADDR_ANY address, maybe that config is indicating brubeck will _listen_ on all local addresses?

AFAICT, using `0.0.0.0` as a synonym for localhost (`127.0.0.1`) (probably) only works on Linux. This doesn't  comply with any standard, since connecting or sending a packet to `INADDR_ANY` doesn't make any sense in the IP protocol.

I would have never thought to try `0.0.0.0` and I couldn't find where or why or how Linux translates this to `localhost`.

It does mean that a literal `0` can be used as a shortcut for `localhost` though!

```
$ ping 0
PING 0 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.084 ms
64 bytes from 127.0.0.1: icmp_seq=2 ttl=64 time=0.050 ms
^C
```

Just thinking of the time I've wasted typing l-o-c-a-l-h-o-s-t in my life gives me shivers.

Reply:
BTW, I would definitely use `127.0.0.1` for localhost in these examples. Use of `0.0.0.0` is sure to confuse somebody (well it did!) :-)

Reply:
By the way @garrettbslone - I should have mentioned this earlier, but a guy I went to school with write an excellent tutorial on socket programming:

https://beej.us/guide/bgnet/html/

Also author of the internet pizza server (try the garlick topping!)

https://beej.us/pizza/